include/irr/asset/IDescriptorSetLayout.h

```diff
@@ -24,20 +25,62 @@ enum E_DESCRIPTOR_TYPE
 
 class IDescriptorSetLayout
 {
-public:
-    struct SBinding
+    template<typename SamplersType>
+    struct SBinding_
     {
         E_DESCRIPTOR_TYPE type;
         uint32_t count;
         E_SHADER_STAGE stageFlags;
-        core::smart_refctd_dynamic_array<SSamplerParams> samplers;
+        SamplersType samplers;
     };
+    using SBinding_internal = SBinding_<const SSamplerParams*>;
+public:
+    using SBinding = SBinding_<core::smart_refctd_dynamic_array<SSamplerParams>>;
 
 protected:
-    IDescriptorSetLayout(core::smart_refctd_dynamic_array<SBinding>&& _bindings) : m_bindings(std::move(_bindings)) {}
+    IDescriptorSetLayout(const SBinding* const _begin, const SBinding* const _end) : 
+        m_bindings(core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<SBinding_internal>>(_end-_begin,nullptr))
+    {
+        size_t bndCount = _end-_begin;
+        size_t immSamplerCount = 0ull;
+        for (size_t i = 0ull; i < bndCount; ++i) {
+            const auto& bnd = _begin[i];
```

Comment:
you can already do the copy of attributes `type`,`count`,`stageFlags` in this loop and compute the first half of the offset as well.